standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/RetryingTxnHandler.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.hadoop.hive.metastore.txn.retryhandling.DataSourceWrapper;
+import org.apache.hadoop.hive.metastore.txn.retryhandling.Retry;
+import org.apache.hadoop.hive.metastore.txn.retryhandling.RetryHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.Arrays;
+
+/**
+ * Responsible to use the {@link RetryHandler} to automatically retry the failed {@link TxnStore} methods if annotated with
+ * the {@link Retry} annotation.
+ */
+public class RetryingTxnHandler implements InvocationHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryingTxnHandler.class);
+
+  /**
+   * Gets the proxy interface for the given {@link TxnStore}.
+   * @param realStore The real {@link TxnStore} to proxy.
+   * @param retryHandler Responsible to re-execute the methods in case of failure.
+   * @return Returns the proxy object capable of retrying the failed calls automatically and transparently. 
+   */
+  public static TxnStore getProxy(TxnStore realStore, RetryHandler retryHandler) {
+    RetryingTxnHandler handler = new RetryingTxnHandler(realStore, retryHandler);
+    return (TxnStore) Proxy.newProxyInstance(
+        RetryingTxnHandler.class.getClassLoader(),
+        new Class[]{ TxnStore.class },
+        handler);
+  }
+
+  private final RetryHandler retryHandler;
+  private final TxnStore realStore;
+
+  private RetryingTxnHandler(TxnStore realStore, RetryHandler retryHandler) {
+    this.realStore = realStore;
+    this.retryHandler = retryHandler;
+  }
+
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    String strArgs = args == null ? "" : StringUtils.join(Arrays.asList(args), ",");
+    String callerId = method.getName() + "(" + strArgs + ")";
+    Retry retry = method.getAnnotation(Retry.class);
+    if (retry != null) {
+      RetryCallProperties properties = new RetryCallProperties()
+          .withCallerId(callerId)
+          .withDataSource(retry.datasource())
+          .withLockInternally(retry.lockInternally())
+          .withRollbackOnError(retry.rollbackOnError())
+          .withRetryOnDuplicateKey(retry.retryOnDuplicateKey())
+          .withTransactionIsolationLevel(retry.transactionIsolation());
+      return retryHandler.executeWithRetry(properties,
+          (DataSourceWrapper dataSourceWrapper) -> {
+            try {
+              LOG.info("Invoking method within retry context: {}", callerId);
+              return method.invoke(realStore, args);
+            } catch (IllegalAccessException | InvocationTargetException | UndeclaredThrowableException e) {
+              if (e.getCause() instanceof MetaException) {
+                throw (MetaException) e.getCause();
```

Comment:
what is the purpose of casting here? same for Runtime exceptiion. check same error handling for non-retry

Reply:
In the retry case we are inside the TransactionalFunction functional interface method which has a signature of `Result execute(DataSourceWrapper dataSourceWrapper) throws SQLException, MetaException;` Since e.getCause() returns Throwable it cannot be re-thrown directly. As a result we need 3 branches:
- the cause was MetaException which is checked so cannot be handled by the runtime branch, we need to cast it
- The cause was RuntimeException, we need to cast it 
- in every other case we wrap it into a RuntimeException (not rethrowing the original).